### PR TITLE
Align rfdfetr preproc/postproc to rf-detr-internal

### PR DESCRIPTION
**Pull request by leeclemnet**
## What does this PR do?

The goal of this PR is to align RF-DETR inference behaviour as closely as possible to rf-detr-internal / roboflow-train behaviour. By doing this we 1) minimize divergence in post-training model-eval metrics vs training-time metrics; 2) ensure deployed RF-DETR models are performing at full accuracy.

Two main divergences identified and resolved by this PR:

#### post-processing divergence
- inference_models rfdetr OD postproc for onnx/trt backends differed from pytorch and rf-detr-internal training.
- rf-detr-internal / OD-torch / IS-all does TopK across flattened (Q, C) scores
- OD-onnx / OD-TRT was doing per-query argmax over (Q, C)
- Practical impact: different detection set when >1 viable detection per query.

#### pre-processing divergence
- inference_models shared preproc follows ultralytics yolo convention: cv2.resize(uint8) → /255 → normalize
- rf-detr-internal uses F.to_tensor → F.normalize → F.resize (all fp32). 
- Difference in resize implementation is enough to change the prediction set - effect is larger for larger input resolution - disappears given image pre-resized to native model resolution
- Numerical difference from order of operations is enough to change the prediction set at low confidence thresholds even when given an image pre-resized to native model resolution

Both issues have larger impact at low confidence thresholds where small numerical differences can noticeably impact the detection set when multiple low-confidence predictions have similar scores. This makes a difference when computing PR-curves, mAP and confidence sweeps. Fixing both issues aligns inference_models to rf-detr-internal to a high precision and almost entirely eliminates the divergence in model-eval vs train P/R/mAP.

## Type of Change

- Bug fix (non-breaking change that fixes an issue)

## Testing

- [x] I have tested this change locally
- [x] I have added/updated tests for this change

**Test details:**

### 1. rf-detr-internal predict() vs inference_models infer() equivalency

Establish equivalency rf-detr-internal <-> torch <-> onnx <-> trt at fp32. TRT packages locally built from onnx for T4.

- Object Detection: https://app.roboflow.one/paul-guerrie/coin-counting/1451
- Instance Segmentation: https://app.roboflow.one/paul-guerrie/asl-instance-seg/159

Both models trained with https://github.com/roboflow/roboflow-train/pull/786 (fp32 eval + onnx export bugfix).

Prediction set for an image is considered a match if all detections can be Hungarian-matched 1:1 with rf-detr-internal detections.

#### Coin-counting coin-counting/1451 — Object Detection — 20 images (12 valid + 8 test) 

vs. rf-detr-internal (torch)

backend | th=0.001 | max \|Δscore\| | max \|Δbox\| | th=0.5 | max \|Δscore\| | max \|Δbox\| | th=0.9
-- | -- | -- | -- | -- | -- | -- | --
torch | 18/20¹ | 7.72e-06 | 0.50 px | 20/20 | 8.71e-04 | 0.49 px | 20/20
onnx | 17/20¹ | 8.30e-06 | 0.50 px | 20/20 | 8.71e-04 | 0.49 px | 20/20
trt | 17/20¹ | 9.74e-06 | 0.51 px | 20/20 | 8.71e-04 | 0.49 px | 20/20

¹ Some residual mismatches due to numerical noise at very low confidence thresholds.

**mAP@50 (model-eval)**
split / class | train | torch_me | onnx_me
-- | -- | -- | --
valid (all) | 0.7772 | 0.7769 | 0.7769
valid · dime | 0.9175 | 0.9180 | 0.9180
valid · nickel | 0.6888 | 0.6888 | 0.6888
valid · penny | 0.7252 | 0.7239 | 0.7239
test (all) | 0.4847 | 0.4852 | 0.4852
test · dime | 0.4446 | 0.4467 | 0.4467
test · nickel | 0.2000 | 0.2000 | 0.2000
test · nickel2 | 0.5545 | 0.5545 | 0.5545
test · penny | 0.7397 | 0.7397 | 0.7397

**P / R @ optimal F1 (model-eval)**
split | source | best_thr | F1 | precision | recall
-- | -- | -- | -- | -- | --
valid | train | — | 0.6562 | 0.5638 | 0.7849
valid | torch_me | 0.4100 | 0.5360 | 0.6050 | 0.7590
valid | onnx_me | 0.4100 | 0.5360 | 0.6050 | 0.7590
test | train | — | 0.5669 | 0.4900 | 0.6726
test | torch_me | 0.0900 | 0.4080 | 0.4900 | 0.6730
test | onnx_me | 0.0900 | 0.4080 | 0.4900 | 0.6730

#### ASL Instance Segmentation asl-instance-seg/159 — Instance Segmentation — 166 images (99 valid + 67 test)

Using RLE mask encoding. vs. rf-detr-internal (torch).

backend | th=0.001 | max \|Δscore\| | max \|Δbox\| | min mask IoU | th=0.5 | max \|Δscore\| | max \|Δbox\| | min mask IoU | th=0.9
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
torch | 165/166¹ | 4.32e-05 | 0.54 px | 0.990 | 166/166 | 5.78e-06 | 0.50 px | 1.0000 | 166/166
onnx | 164/166¹ | 4.20e-05 | 0.54 px | 0.986 | 166/166 | 1.93e-04 | 0.50 px | 0.9999 | 166/166
trt | 165/166¹ | 2.60e-05 | 0.54 px | 0.994 | 166/166 | 5.83e-05 | 0.50 px | 1.0000 | 166/166

¹ Some residual mismatches due to numerical noise at very low confidence thresholds.

**mAP@50 (model-eval)**

split / class | train | torch_me | onnx_me | Δ (o−t)
-- | -- | -- | -- | --
valid (all) | 0.9786 | 0.9775 | 0.9775 | 0.0000
valid · F | 0.8754 | 0.8678 | 0.8678 | 0.0000
valid · J | 0.9670 | 0.9670 | 0.9670 | 0.0000
valid · M | 0.8350 | 0.8350 | 0.8350 | 0.0000
valid · N | 0.9670 | 0.9670 | 0.9670 | 0.0000
valid · P | 0.8203 | 0.8020 | 0.8020 | 0.0000
(other 20 valid classes) | 1.0000 | 1.0000 | 1.0000 | 0.0000
test (all) | 0.9911 | 0.9911 | 0.9911 | 0.0000
test · E | 0.8350 | 0.8350 | 0.8350 | 0.0000
test · T | 0.9505 | 0.9505 | 0.9505 | 0.0000
(other 22 test classes) | 1.0000 | 1.0000 | 1.0000 | 0.0000

**P / R @ optimal F1 (model-eval)**

split | source | best_thr | F1 | precision | recall
-- | -- | -- | -- | -- | --
valid | train | — | 0.9252 | 0.9179 | 0.9327
valid | torch_me | 0.6000 | 0.9230 | 0.9260 | 0.9330
valid | onnx_me | 0.6000 | 0.9230 | 0.9260 | 0.9330
test | train | — | 0.9541 | 0.9464 | 0.9618
test | torch_me | 0.4300 | 0.9410 | 0.9460 | 0.9620
test | onnx_me | 0.4300 | 0.9410 | 0.9460 | 0.9620

### 2. Update rf-detr regression test expected values

Torch+ONNX+TRT should produce numerically close results to each other and rf-detr-internal at fp32.

Method

For each test package: run inference_models[torch](raw_image) → predictions A. Externally apply the same first-stage pre-processing (single-step: apply_pre_processing_to_numpy_image; two-step: _dataset_version_resize_uint8) to produce an intermediate image. Run rfdetr_internal.predict(intermediate_image) with resolution=training_input_size → predictions B. Translate B back into A's coordinate space (single-step: + static_crop_offset; two-step: inverse letterbox via the metadata's scale_width/scale_height + pad info, then + static_crop_offset). Compare A vs B.

Results

package | needs_two_step | dets | max \|Δscore\| | max \|Δbox\| | cls match | verdict
-- | -- | -- | -- | -- | -- | --
coin-counting-rfdetr-nano-torch-cs-stretch-640 | False | 10 | 1.01e-06 | 0.49 px | 10/10 | MATCH
rfdetr-nano-torch-static-crop-letterbox-640 | False | 8 | 5.96e-07 | 0.48 px | 8/8 | MATCH
rfdetr-nano-torch-center-crop-640 | False | 9 | 1.19e-07 | 0.49 px | 9/9 | MATCH
rfdetr-nano-torch-static-crop-center-crop-640 | False | 8 | 8.34e-07 | 0.49 px | 8/8 | MATCH
rfdetr-nano-torch-static-bs-nonsquare-letterbox | True | 9 | 1.79e-07 | 0.44 px | 9/9 | MATCH
rfdetr-seg-torch-stretch | False | 1 | 1.28e-05 | 0.36 px | 1/1 | MATCH
rfdetr-seg-torch-letterbox | False | 1 | 1.19e-07 | 0.48 px | 1/1 | MATCH
rfdetr-seg-torch-center-crop | False | 1 | 0.0e+00 | 0.40 px | 1/1 | MATCH
rfdetr-seg-torch-static-crop-stretch | False | 1 | 1.19e-07 | 0.46 px | 1/1 | MATCH
rfdetr-seg-torch-static-crop-letterbox | False | 1 | 0.0e+00 | 0.19 px | 1/1 | MATCH
rfdetr-seg-torch-static-crop-center-crop | False | 1 | 2.74e-06 | 0.41 px | 1/1 | MATCH
rfdetr-seg-torch-static-bs-nonsquare-letterbox | True | 1 | 2.38e-07 | 0.32 px | 1/1 | MATCH

Note: TRT regression tests are for different model than the Torch/ONNX ones, so didn't compare against rf-detr-internal directly. Will come back and align the TRT tests to the other two in a followup PR.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have updated the documentation accordingly (if applicable)

### CODE REVIEW

File: inference_models/tests/integration_tests/models/test_rfdetr_predictions_onnx.py [hunk 1]
```diff
@@ -26,7 +26,19 @@ def test_onnx_package_with_stretch_resize_and_contrast_stretching_numpy(
     assert torch.allclose(
         predictions[0].confidence.cpu(),
         torch.tensor(
-            [0.8575, 0.8568, 0.8105, 0.7940, 0.7364, 0.6872, 0.6419, 0.5810, 0.5425]
+            [
```
Comment (leeclemnet) [inline inference_models/tests/integration_tests/models/test_rfdetr_predictions_onnx.py]: Test now matches expected values from torch backend tests
https://github.com/roboflow/inference/blob/main/inference_models/tests/integration_tests/models/test_rfdetr_predictions_torch.py#L62

File: inference/core/models/inference_models_adapters.py [hunk 2]
```diff
@@ -121,6 +121,7 @@ def __init__(self, model_id: str, api_key: str = None, **kwargs):
         self.class_names = list(self._model.class_names)
 
     def map_inference_kwargs(self, kwargs: dict) -> dict:
+        kwargs["input_color_format"] = "bgr"
```
Comment (leeclemnet) [inline inference/core/models/inference_models_adapters.py:124]: This was implicit before. common preprocessing correctly assumes BGR if input_color_mode is None, but it's better to be explicit.

File: inference_models/inference_models/models/rfdetr/pre_processing.py [hunk 3]
```diff
@@ -1,31 +1,46 @@
...
+torch.Tensor inputs (advanced caller, float CHW [0, 1]):
+    tensor F.resize → F.normalize
+"""
+
 from typing import List, Optional, Tuple, Union
 
 import numpy as np
 import torch
+import torchvision.transforms.functional as TF
+from PIL import Image
```
Comment (PawelPeczek-Roboflow) [inline inference_models/inference_models/models/rfdetr/pre_processing.py:19]: do we absolutely need conversion to PIL? I was hoping to keep everything with cv2 and torch.

**Review by PawelPeczek-Roboflow — APPROVED** [review]

**Comment by dkosowski87** [comment]
@leeclemnet did you benchmark the latency change somehow?

**Comment by leeclemnet** [comment]
> @leeclemnet did you benchmark the latency change somehow?

@dkosowski87 the benchmark numbers I quoted to the Codeflash folks here were from this branch. Comparing to what they were seeing on main, the latency impact is small: https://roboflow.slack.com/archives/C08JS1LEMEG/p1777410521218969?thread_ts=1776455846.033309&cid=C08JS1LEMEG

It's a bit of a moot point given the optimized Triton kernels they came up with accelerating rf-detr inference by 5x: https://roboflow.slack.com/archives/C08JS1LEMEG/p1778098988619889

I want to ensure we have a correct reference implementation and tests so we can tweak the custom Triton implementation to match as close as possible.

File: inference_models/inference_models/models/rfdetr/pre_processing.py (hunk 3 above)
Comment (leeclemnet) [reply inference_models/inference_models/models/rfdetr/pre_processing.py:19]: The PIL vs cv2 resize outputs are slightly different, and the scores shift just enough (up to ~5e-3 depending on how much resize) to change the detection set vs rf-detr-internal, especially at low confidence. It doesn't make a big difference to production since this is mostly happening below typical operating points but it does change the low-confidence region of the PR curve which produces slightly different mAP vs. training.

It's a bit of a moot point given the optimized Triton kernels the Codeflash people came up with accelerating rf-detr inference by 5x: https://roboflow.slack.com/archives/C08JS1LEMEG/p1778098988619889. We can use the PIL/numpy reference implementation and tests to tweak their custom Triton implementation so it matches as closely as possible.
